src/system/SystemWakeEvent.cpp

```diff
@@ -66,32 +66,56 @@ Error SystemWakeEvent::Open()
     return CHIP_SYSTEM_NO_ERROR;
 }
 
-void SystemWakeEvent::Close()
+Error SystemWakeEvent::Close()
 {
-    (void) ::close(mFDs[FD_WRITE]);
-    (void) ::close(mFDs[FD_READ]);
+    int res = 0;
+
+    res |= ::close(mFDs[FD_WRITE]);
+    res |= ::close(mFDs[FD_READ]);
     mFDs[FD_READ] = mFDs[FD_WRITE] = -1;
+
+    if (res < 0)
+    {
+        return chip::System::MapErrorPOSIX(errno);
+    }
+
+    return CHIP_SYSTEM_NO_ERROR;
 }
 
-void SystemWakeEvent::Confirm()
+Error SystemWakeEvent::Confirm()
 {
     uint8_t buffer[128];
+    int res;
+
+    do
+    {
+        res = ::read(mFDs[FD_READ], buffer, sizeof(buffer));
+        if (res < 0 && errno != EAGAIN && errno != EWOULDBLOCK)
+        {
+            return chip::System::MapErrorPOSIX(errno);
+        }
+    } while (res == sizeof(buffer));
 
-    while (::read(mFDs[FD_READ], buffer, sizeof(buffer)) == sizeof(buffer))
-        continue;
+    return CHIP_SYSTEM_NO_ERROR;
 }
 
-void SystemWakeEvent::Notify()
+Error SystemWakeEvent::Notify()
 {
     char byte = 1;
-    (void) ::write(mFDs[FD_WRITE], &byte, 1);
+
+    if (::write(mFDs[FD_WRITE], &byte, 1) < 0 && errno != EAGAIN && errno != EWOULDBLOCK)
```

Comment:
Need a loop for the EAGAIN case?

If it can happen, looks like this drops the notification on the floor.

Does Zephyr have TEMP_FAILURE_RETRY ?

Reply:
I don't think it needs a loop. If we get EAGAIN during write, then the event is already set and we don't need to do anything (select will return anyway). During read we are only interested in clearing the event to make the select call block, so EAGAIN is one of the desired outcomes.